### ISSUE TITLE
Fails to detect broken connections in some cases

### ISSUE THREAD

**Issue by TheTincho** [psycopg/psycopg2#263]
Hi,

I have been trying to chase down a bug I am experiencing: sometimes the connection to the database would go away, and neither psycopg2 nor sqlalchemy detect the condition, and thus, the connection is not removed from the connection pool.

It seems to be the same issue as #196 , which is marked as solved, but I can reproduce this behaviour with psycopg 2.5.3, and problems other than the server restarting. In my case it seems that connections are timing out (the error is `SSL SYSCALL error: EOF detected`), but closing manually the filedescriptor triggers the same issue):

```
$ python
Python 2.7.3 (default, Mar 13 2014, 11:03:55) 
[GCC 4.7.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import psycopg2
>>> conn = psycopg2.connect('postgresql://XXX:YYY@localhost/testsrunner')
>>> cur = conn.cursor()
>>> cur.execute("SELECT 1")
>>> cur.fetchall()
[(1,)]
>>> conn.closed
0
>>> conn.poll()
0
>>> conn.closed
0
>>> 
[1]+  Stopped                 python
$ ps
  PID TTY          TIME CMD
14434 pts/3    00:00:00 bash
18299 pts/3    00:00:00 python
18336 pts/3    00:00:00 ps
$ lsof -p 18299|grep TCP
python  18299 tincho    3u  IPv4 13795332      0t0     TCP localhost:46295->localhost:postgresql (ESTABLISHED)
$ fg
python


>>> import os
>>> os.close(3)
>>> cur = conn.cursor()
>>> cur.execute("SELECT 1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
psycopg2.OperationalError: SSL SYSCALL error: Bad file descriptor

>>> conn.closed
0
>>> conn.poll()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
psycopg2.OperationalError: SSL SYSCALL error: Bad file descriptor

>>> conn.closed
0
>>> psycopg2.__version__
'2.5.3 (dt dec mx pq3 ext)'
```

It is worth noting that this does not seem to be an SSL-related issue, as I can reproduce the same problem with a UNIX domain socket:

```
>>> conn = psycopg2.connect('postgresql://')
[..]
$ lsof -p 18427|grep socket
python  18427 postgres    3u  unix 0xffff880000ea43c0      0t0 13797662 socket
$ fg
python


>>> import os
>>> os.close(3)
>>> cur = conn.cursor()
>>> cur.execute("SELECT 1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
psycopg2.OperationalError: could not receive data from server: Bad file descriptor

>>> conn.closed
0
>>> conn.poll()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
psycopg2.OperationalError: could not receive data from server: Bad file descriptor

>>> conn.closed
0
```

**Comment by rkonda** [comment]
what is the resolution for this issue?

**Comment by dvarrazzo** [comment]
That psycopg won't do anything actively such a poll to check the detection is broken. If your application finds a connection in broken/inconsistent state it is your responsibility to dispose of it and open a new one. If you use the connection pool it should do that for you (but on putconn, not on getconn) and if it doesn't it is our bug to fix.

I have recently improved some internal handling quirknesses so some of the issues here could have been solved, but more along the line of returning a better error message, not automatic reconnection or anything too fancy.
